src/main/java/com/spotify/folsom/client/DefaultRawMemcacheClient.java

```diff
@@ -189,43 +187,54 @@ public int getOutstandingRequests() {
   }
 
   @Override
-  public <T> ListenableFuture<T> send(final Request<T> request) {
-    if (!tryIncrementPending()) {
+  public <T> CompletableFuture<T> send(final Request<T> request) {
 
+    if (!tryIncrementPending()) {
       // Do the disconnect check in here instead of outside
       // to get better performance in the happy case.
       String disconnectReason = this.disconnectReason.get();
       if (disconnectReason != null) {
-        MemcacheClosedException exception = new MemcacheClosedException(disconnectReason);
-        return onExecutor(Futures.<T>immediateFailedFuture(exception));
+        return newExceptionallyCompletedFuture(new MemcacheClosedException(disconnectReason));
       }
 
-      return onExecutor(Futures.<T>immediateFailedFuture(new MemcacheOverloadedException(
-              "too many outstanding requests")));
+      return newExceptionallyCompletedFuture(new MemcacheOverloadedException(
+              "too many outstanding requests"));
     }
+
     if (request instanceof SetRequest) {
       SetRequest setRequest = (SetRequest) request;
       byte[] value = setRequest.getValue();
       if (value.length > maxSetLength) {
-        return (ListenableFuture<T>) onExecutor(
-                Futures.immediateFuture(MemcacheStatus.VALUE_TOO_LARGE));
+        // Is this cast correct. Shouldn't return type be CompletableFuture<MemcacheStatus>?
```

Comment:
Good question, I don't remember why that cast was needed. Will need to investigate.